wire.go

```diff
@@ -83,21 +119,79 @@ func (cdc *Codec) MarshalBinary(o interface{}) ([]byte, error) {
 		}
 	}
 
-	w := new(bytes.Buffer)
+	// Encode Wire:binary bytes.
+	var bz []byte
+	buf := new(bytes.Buffer)
 	rt := rv.Type()
 	info, err := cdc.getTypeInfo_wlock(rt)
 	if err != nil {
 		return nil, err
 	}
-	err = cdc.encodeReflectBinary(w, info, rv, FieldOptions{})
+	err = cdc.encodeReflectBinary(buf, info, rv, FieldOptions{})
 	if err != nil {
 		return nil, err
 	}
-	return w.Bytes(), nil
+	bz = buf.Bytes()
+
+	return bz, nil
 }
 
+// Like UnmarshalBinaryBare, but will first decode the byte-length prefix.
 // UnmarshalBinary will panic if ptr is a nil-pointer.
+// Returns an error if not all of bz is consumed.
 func (cdc *Codec) UnmarshalBinary(bz []byte, ptr interface{}) error {
+
+	// Read byte-length prefix.
+	u64, n := binary.Uvarint(bz)
+	if n < 0 {
+		return fmt.Errorf("Error reading msg byte-length prefix: got code %v", n)
+	}
+	if u64 > uint64(len(bz)-n) {
+		return fmt.Errorf("Not enough bytes to read in UnmarshalBinary, want %v more bytes but only have %v",
+			u64, len(bz)-n)
+	} else if u64 < uint64(len(bz)-n) {
+		return fmt.Errorf("Bytes left over in UnmarshalBinary, should read %v more bytes but have %v",
+			u64, len(bz)-n)
+	}
+	bz = bz[n:]
+
+	// Decode.
+	return cdc.UnmarshalBinaryBare(bz, ptr)
+}
+
+// Like UnmarshalBinaryBare, but will first read the byte-length prefix.
+// UnmarshalBinaryReader will panic if ptr is a nil-pointer.
+// If maxSize is 0, there is no limit (not recommended).
+func (cdc *Codec) UnmarshalBinaryReader(r io.Reader, ptr interface{}, maxSize int64) error {
+	if maxSize < 0 {
```

Comment:
Neither can it be 0 otherwise we are back to square 1 where we can read unbounded amounts of memory.

How come are even passing in a maxSize parameter?